### PR TITLE
Change the representation of reachability claims, part 1

### PR DESCRIPTION
**Pull request by ana-pantilie**
Part of https://github.com/kframework/kore/issues/1278

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/test/Test/Kore/Step/Rule/Common.hs [hunk 1]
```diff
@@ -0,0 +1,114 @@
+module Test.Kore.Step.Rule.Common
+    ( Pair (..)
+    , RuleBase (..)
+    ) where
+
+import Prelude.Kore
+
+
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Rule/Common.hs:8]:
```suggestion
```

File: kore/test/Test/Kore/Step/Rule/Common.hs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import Kore.Step.RulePattern
+    ( RHS (RHS)
+    , RewriteRule (RewriteRule)
+    , RulePattern (RulePattern)
+    )
+import qualified Kore.Step.RulePattern as OLD
+
+newtype Pair variable = Pair (TermLike variable, Predicate variable)
+
+class RuleBase base rule where
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Rule/Common.hs:40]: Is this class very useful, now that we can have whole `Pattern`s on the left- and right-hand sides?

File: kore/src/Kore/Internal/Conditional.hs [hunk 3]
```diff
@@ -245,6 +248,17 @@ instance
   where
     from = pure
 
+instance
+    InternalVariable variable
+    => From (Map (SomeVariable variable) (TermLike variable)) (Conditional variable ())
+  where
+    from =
+        from @(Predicate variable) @(Conditional variable ())
+        . from @(Substitution variable) @(Predicate variable)
```
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply kore/src/Kore/Internal/Conditional.hs:257]: Yes, this isn't right. Thanks for spotting this!

File: kore/src/Kore/Step/ClaimPattern.hs [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+instance SOP.Generic AllPathRule
+
+instance SOP.HasDatatypeInfo AllPathRule
+
+instance Debug AllPathRule
+
+instance Diff AllPathRule
+
+instance Unparse AllPathRule where
+    unparse claimPattern' =
```
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply kore/src/Kore/Step/ClaimPattern.hs:360]: I actually found this very strange when copying the instance from `RulePattern`. I thought claims should unparse to their pattern form. Adding the `claim {}` declaration at the beginning would make this a sentence.

File: kore/test/Test/Kore/Step/Rule/Common.hs (hunk 2 above)
Comment (ana-pantilie) [reply kore/test/Test/Kore/Step/Rule/Common.hs:40]: The tests in `Simplify` should work on `RewriteRule`s as well. I might remove this class once I clean up all the old code, but right now it seemed the cleanest way to run the tests with old/new claims and rewrite rules.

File: kore/src/Kore/Step/ClaimPattern.hs [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+-- 'TermLike' is parsed into an 'OrPattern' and a list of
+-- existentially quantified variables.
+claimPattern
+    :: Pattern RewritingVariableName
+    -> TermLike RewritingVariableName
+    -> ClaimPattern
+claimPattern left rightTerm =
+    ClaimPattern
+        { left
+        , right = OrPattern.parseFromTermLike rightTerm
```
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply kore/src/Kore/Step/ClaimPattern.hs:178]: I gave this a little more thought, and maybe we won't actually need to parse the RHS in any other places except when actually parsing the definition. I'll remove it now.

**Pull request merged**